schematic/schemas/data_model_json_schema.py

```diff
@@ -0,0 +1,356 @@
+		"""Add a list of nodes to the "enum" key in a given JSON schema object.
+
+		Args:
+			node_name: Name of the "main" / "head" key in the JSON schema / object.
+			node_range: List of nodes to be added to the JSON object.
+			blank: If True, add empty node to end of node list.
+				   If False, do not add empty node to end of node list.
+
+		Returns:
+			JSON object with nodes.
+		"""
+		if blank:
+			schema_node_range = {node_name: {"enum": node_range + [""]}}
+		else:
+			schema_node_range = {node_name: {"enum": node_range}}
+
+		return schema_node_range
+
+	def get_json_validation_schema(self, source_node: str, schema_name: str) -> Dict:
+		'''
+		A refactor of get_json_schema_requirements() from the
+		schema generator.
+		Consolidated method that aims to gather dependencies and value constraints across terms / nodes in a schema.org schema and store them in a jsonschema /JSON Schema schema.
+
+		It does so for any given node in the schema.org schema (recursively) using the given node as starting point in the following manner:
+		1) Find all the nodes / terms this node depends on (which are required as "additional metadata" given this node is "required").
+		2) Find all the allowable metadata values / nodes that can be assigned to a particular node (if such a constraint is specified on the schema).
+
+		Args:
+			source_node: Node from which we can start recursive dependancy traversal (as mentioned above).
+			schema_name: Name assigned to JSON-LD schema (to uniquely identify it via URI when it is hosted on the Internet).
+
+		Returns:
+			JSON Schema as a dictionary.
+		'''
+		json_schema = {
+			"$schema": "http://json-schema.org/draft-07/schema#",
+			"$id": "http://example.com/" + schema_name,
+			"title": schema_name,
+			"type": "object",
+			"properties": {},
+			"required": [],
+			"allOf": [],
+		}
+
+		nodes_to_process = (
+			[]
+		)  # list of nodes to be checked for dependencies, starting with the source node
+		processed_nodes = (
+			[]
+		)  # keep of track of nodes whose dependencies have been processed
+		reverse_dependencies = (
+			{}
+		)  # maintain a map between conditional nodes and their dependencies (reversed) -- {dependency : conditional_node}
+		range_domain_map = (
+			{}
+		)  # maintain a map between range nodes and their domain nodes {range_value : domain_value}
+		# the domain node is very likely the parentof ("parentOf" relationship) of the range node
+
+		root_dependencies = self.DME.get_adjacent_nodes_by_relationship(
+			node_label = source_node, relationship=self.rel_dict['requiresDependency']['edge_key'],
+		)
+
+		# if root_dependencies is empty it means that a class with name 'source_node' exists
+		# in the schema, but it is not a valid component
+		if not root_dependencies:
+			raise ValueError(f"'{source_node}' is not a valid component in the schema.")
+
+		nodes_to_process += root_dependencies
+
+		process_node = nodes_to_process.pop(0)
+
+		while process_node:
```

Comment:
This while loop is a bit out of control? If I had to come in and work on it , I wouldn't know where to start. The comments are helpful, but there is no way I could keep track of how this whole loop worked in my head. I strongly suggest breaking this up into more managable chunks.

Reply:
@andrewelamb Yes, we decided to leave this outside of the scope of this refactor.